### ISSUE TITLE
Dictionary Associations can be directly modified, leading to bizarre KeyNotFound errors

### ISSUE THREAD

**Issue by sdasda7777** [guillep/pharo-core#13353]
**Bug description**
I was seriously confused for the last two hours or so, because every time I did `dict at: string` on my Dictionaries, I would get a `KeyNotFound` error. However I could see the key was in there, and even `dict keys includes: string` would come back as True. Following are pictures that show the two strings byte per byte.

![image](https://user-images.githubusercontent.com/17746796/230686432-467da7a1-b6c9-4c6b-9c59-58d4fccced37.png)

![image](https://user-images.githubusercontent.com/17746796/230686458-f9ee7795-edd8-40c1-a058-724e018724aa.png)

The reason was that I was doing `dict associationsDo: [ :pair | pair key: (fun (pair key))]`, which changed the key, but I assume it didn't change where it was in the underlying tree structure, or something like that.

In short, I don't think keys should be exposed to modification in this way. I would have never figured it out if I didn't have some knowlege of how Dictionaries usually work, which I think common person doesn't. If you need to know how abstraction works under the hood to use it properly, that's a bad abstraction.

**To Reproduce**
Create a dictionary, iterate associations, directly modify some keys. The tree will be built on outdated information, and you will only find anything if you're very lucky.

**Comment by svenvc** [comment]
Hi,

Thanks for the feedback.

You are correct but it has always been like that and it unlikely to change soon.

Yes, associations are part of the internal structure of the dictionary and should never be changed in place.

Maybe the key could be made readonly once the association is created, I am not sure that is possible at that granularity.

Sven
